Mathlib/LinearAlgebra/FreeModule/Finite/Rank.lean

```diff
@@ -227,3 +232,496 @@ theorem Submodule.finrank_le_finrank_of_le {s t : Submodule R M} [Module.Finite
+
+variable (R V)
+
+instance Module.finite_bot : Module.Finite R (⊥ : Submodule R V) := inferInstance
+
+variable {R V}
+
+theorem Submodule.bot_eq_top_of_rank_eq_zero [NoZeroSMulDivisors R V] (h : Module.rank R V = 0) :
+    (⊥ : Submodule R V) = ⊤ := by
+  rw [rank_zero_iff] at h
+  exact Subsingleton.elim _ _
+#align bot_eq_top_of_rank_eq_zero Submodule.bot_eq_top_of_rank_eq_zero
+
+@[simp]
+theorem Submodule.rank_eq_zero [NoZeroSMulDivisors R V] {S : Submodule R V} :
+    Module.rank R S = 0 ↔ S = ⊥ :=
+  ⟨fun h =>
+    (Submodule.eq_bot_iff _).2 fun x hx =>
+      congr_arg Subtype.val <|
+        ((Submodule.eq_bot_iff _).1 <| Eq.symm <| Submodule.bot_eq_top_of_rank_eq_zero h) ⟨x, hx⟩
+          Submodule.mem_top,
+    fun h => by rw [h, rank_bot]⟩
+#align rank_eq_zero Submodule.rank_eq_zero
+
+@[simp]
+theorem Submodule.finrank_eq_zero [NoZeroSMulDivisors R V] {S : Submodule R V} [Module.Finite R S] :
+    finrank R S = 0 ↔ S = ⊥ := by
+  rw [← Submodule.rank_eq_zero, ← finrank_eq_rank, ← @Nat.cast_zero Cardinal, Cardinal.natCast_inj]
+#align finrank_eq_zero Submodule.finrank_eq_zero
+
+end ZeroRank
+
+namespace Submodule
+
+open IsNoetherian FiniteDimensional
+
+variable [AddCommGroup V] [Ring R] [StrongRankCondition R] [Module R V]
+
+theorem fg_iff_finite (s : Submodule R V) : s.FG ↔ Module.Finite R s :=
+  (finite_def.trans (fg_top s)).symm
+
+/-- The sup of two fg submodules is finite. Also see `Submodule.FG.sup`. -/
+instance finite_sup (S₁ S₂ : Submodule R V) [h₁ : Module.Finite R S₁]
+    [h₂ : Module.Finite R S₂] : Module.Finite R (S₁ ⊔ S₂ : Submodule R V) := by
+  rw [finite_def] at *
+  exact (fg_top _).2 (((fg_top S₁).1 h₁).sup ((fg_top S₂).1 h₂))
+
+/-- The submodule generated by a finite supremum of finite dimensional submodules is
+finite-dimensional.
+
+Note that strictly this only needs `∀ i ∈ s, FiniteDimensional K (S i)`, but that doesn't
+work well with typeclass search. -/
+instance finite_finset_sup {ι : Type*} (s : Finset ι) (S : ι → Submodule R V)
+    [∀ i, Module.Finite R (S i)] : Module.Finite R (s.sup S : Submodule R V) := by
+  refine'
+    @Finset.sup_induction _ _ _ _ s S (fun i => Module.Finite R ↑i) (Module.finite_bot R V)
+      _ fun i _ => by infer_instance
+  · intro S₁ hS₁ S₂ hS₂
+    exact Submodule.finite_sup S₁ S₂
+
+/-- The submodule generated by a supremum of finite dimensional submodules, indexed by a finite
+sort is finite-dimensional. -/
+instance finite_iSup {ι : Sort*} [Finite ι] (S : ι → Submodule R V)
+    [∀ i, Module.Finite R (S i)] : Module.Finite R ↑(⨆ i, S i) := by
+  cases nonempty_fintype (PLift ι)
+  rw [← iSup_plift_down, ← Finset.sup_univ_eq_iSup]
+  exact Submodule.finite_finset_sup _ _
+
+end Submodule
+
+section
+
+variable [Ring R] [StrongRankCondition R] [AddCommGroup V] [Module R V]
+
+instance Module.Finite.finsupp {ι : Type*} [_root_.Finite ι] [Module.Finite R V] :
+    Module.Finite R (ι →₀ V) :=
+  Module.Finite.equiv (Finsupp.linearEquivFunOnFinite R V ι).symm
```

Comment:
Can you just put these into RingTheory.Finiteness? They don't require StrongRankCondition at all.

Reply:
A lot of lemmas in this file are already misplaced and will need a cleanup PR later. I can move these lemmas first if you insist.

Reply:
As [mathlib docs shows]([Submodule.fg_iff_finite](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FreeModule/Finite/Rank.html#Submodule.fg_iff_finite)) these already doesn't use StrongRankCondition. I removed it from `variable` in #9152 so that it won't be forgotten. I refrained from moving them, awaiting the misplacement cleanup PR.